src/librustc_typeck/collect.rs

```diff
@@ -2877,9 +2877,16 @@ fn codegen_fn_attrs(tcx: TyCtxt<'_>, id: DefId) -> CodegenFnAttrs {
         } else if attr.check_name(sym::thread_local) {
             codegen_fn_attrs.flags |= CodegenFnAttrFlags::THREAD_LOCAL;
         } else if attr.check_name(sym::track_caller) {
-            if tcx.fn_sig(id).abi() != abi::Abi::Rust {
-                struct_span_err!(tcx.sess, attr.span, E0737, "`#[track_caller]` requires Rust ABI")
+            if !tcx.is_closure(id) {
+                if tcx.fn_sig(id).abi() != abi::Abi::Rust {
```

Comment:
Closures are not `Abi::Rust`, so the error should likely be reported for them as well.

Reply:
Addressed.